cvmfs/sync_union_tarball.h

```diff
@@ -11,40 +11,63 @@
 
 #include "sync_union.h"
 
+#include <pthread.h>
+
+#include <map>
 #include <set>
 #include <string>
 
+#include "duplex_libarchive.h"
+#include "sync_mediator.h"
+#include "util_concurrency.h"
+
 namespace publish {
 
 class SyncUnionTarball : public SyncUnion {
  public:
-  SyncUnionTarball(SyncMediator *mediator, const std::string &tarball_path,
-                   const std::string &base_directory);
+  SyncUnionTarball(AbstractSyncMediator *mediator,
+                   const std::string &rdonly_path,
+                   const std::string &tarball_path,
+                   const std::string &base_directory,
+                   const std::string &to_delete);
+
+  ~SyncUnionTarball();
 
   /*
    * Check that the tarball is actually valid and that can be open.
-   * Similarly it creates the base directory or check that it has the ability to
-   * write in it.
-   * Finally it untar the tarball, recursively.
    */
   bool Initialize();
 
   /*
-   * Simply create the SyncItem from the basedirectory opened
+   * We start by deleting the entity that we are request to delete.
+   * Then we move on to extracting the tarball.
+   * For each directory we found we remember it associated with its SyncItem on
+   * the `dirs_` map.
+   * Similarly we remember where nested catalogs should be placed in
+   * `to_create_catalog_dirs_`.
+   * After we finish to uncompress the tarball we come back to iterate over
+   * `to_create_catalog_dirs_` and we created the nested catalogs.
    */
   void Traverse();
 
+  std::string UnwindWhiteoutFilename(SharedPtr<SyncItem> entry) const;
+  bool IsOpaqueDirectory(SharedPtr<SyncItem> directory) const;
+  bool IsWhiteoutEntry(SharedPtr<SyncItem> entry) const;
+
  private:
+  struct archive *src;
   const std::string tarball_path_;
   const std::string base_directory_;
-  std::set<std::string> to_recur_;
+  const std::string to_delete_; /* entity to delete */
+  std::set<std::string>
+      know_directories_; /* directory that we know already exist */
```

Comment:
Doxygen style comments.  The comment should explain a bit more, namely the problem of random order of contents in the tarball which makes it necessary to fix-up directories later on.